The following questions concern a small snapshot of top-of-book data in futures markets
and ask you to consider implied orders and trades. Without showing all contracts that imply
into these markets and without showing the market-by-order data and knowing the matching algorithms,
it is impossible to say with certainty what limit orders get filled. Instead, these questions
ask whether it is possible to get filled. You do not need to determine actual fills.

You should show your work, but it does not need to be code. You _can_ code a solution, but you
could simply present your answer in a markdown cell or write out your solutions in another document.

In [1]:
import pandas as pd

top_of_book = pd.DataFrame(
    [
        {
            "symbol": "SR3H6",
            "bid_sz_00": 431,
            "bid_px_00": 96.6,
            "ask_px_00": 96.605,
            "ask_sz_00": 2190,
        },
        {
            "symbol": "SR3U7",
            "bid_sz_00": 5563,
            "bid_px_00": 96.96,
            "ask_px_00": 96.965,
            "ask_sz_00": 52,
        },
        {
            "symbol": "SR3H6-SR3U7",
            "bid_sz_00": 7,
            "bid_px_00": -37,
            "ask_px_00": -36,
            "ask_sz_00": 7,
        },
    ]
)
top_of_book

,symbol,bid_sz_00,bid_px_00,ask_px_00,ask_sz_00
0,SR3H6,431,96.60,96.605,2190
1,SR3U7,5563,96.96,96.965,52
2,SR3H6-SR3U7,7,-37.00,-36.000,7


1. What are the bid and ask, price and size, in the spread market implied IN by the outright legs?

Note that the spread market prices are 100 times the legs' prices. In fact, it is not uncommon to see
the legs in these units as  well. In the solution, I will do everything in the units of the spreads.

Best implied bid price = SR3H6 Bid - SR3U7 Ask = 9660- 9696.5 = -36.5

Implied size at the best implied bid price = min(SR3H6 Bid Size, SR3U7 Ask Size) = min(431, 52) = 52

Best implied ask price = SR3H6 Ask - SR3U7 Bid = 9660.5 - 9696 = -35.5

Implied size at the best implied ask price = min(SR3H6 Ask Size, SR3U7 Bid Size) = min(2190, 5563) = 2190



2. If a market order sells 5 spreads, what price will trade? Could it be implied in the legs?

In part 1, we found that -36.5 is the best bid price, and it is implied so the 5 spread order will
trade against those first, and with 52 implied bid size, that will fill the market order.

3. If a market order buys 5 spreads, what price will trade? Could it be implied in the legs?

The implied offer of -35.5 is not at the top of the book, and the 7 real offers will fill the
market order at -36. No implied orders will trade.

In [2]:
top_of_book = pd.DataFrame(
    [
        {
            "symbol": "ZSX5",
            "bid_sz_00": 23,
            "bid_px_00": 1008.5,
            "ask_px_00": 1008.75,
            "ask_sz_00": 46,
        },
        {
            "symbol": "ZSF6",
            "bid_sz_00": 72,
            "bid_px_00": 1026.25,
            "ask_px_00": 1026.5,
            "ask_sz_00": 84,
        },
        {
            "symbol": "ZSH6",
            "bid_sz_00": 16,
            "bid_px_00": 1041.75,
            "ask_px_00": 1042,
            "ask_sz_00": 8,
        },
        {
            "symbol": "ZSX5-ZSF6",
            "bid_sz_00": 4141,
            "bid_px_00": -18,
            "ask_px_00": -17.75,
            "ask_sz_00": 2187,
        },
        {
            "symbol": "ZSF6-ZSH6",
            "bid_sz_00": 348,
            "bid_px_00": -15.5,
            "ask_px_00": -15.25,
            "ask_sz_00": 1167,
        },
    ]
)
top_of_book

,symbol,bid_sz_00,bid_px_00,ask_px_00,ask_sz_00
0,ZSX5,23,1008.50,1008.75,46
1,ZSF6,72,1026.25,1026.50,84
2,ZSH6,16,1041.75,1042.00,8
3,ZSX5-ZSF6,4141,-18.00,-17.75,2187
4,ZSF6-ZSH6,348,-15.50,-15.25,1167


4. What is the best bid and offer, price and size,
implied OUT in the `ZSF6` market by each of the spreads and other legs?

There are two combinations to imply `ZSF6` from the other markets.

1. `ZSX5` and `ZSX5-ZSF6`.

Implied bid from bidding `ZSX5` and offering `ZSX5-ZSF6`: 1008.50 - -17.75 = 1026.25

Implied bid size = min(23, 2187) = 23

Implied offer from offering `ZSX5` and bidding `ZSX5-ZSF6`: 1008.75 - -18 = 1026.75

Implied offer size = min(46, 4141) = 46

2. `ZSH6` and  `ZSF6-ZSH6`.

Implied bid from bidding `ZSF6-ZSH6` and bidding `ZSH6`: -15.5 + 1041.75 = 1026.25

Implied bid size = min(348, 16) = 16

Implied offer from offering `ZSF6-ZSH6` and offering `ZSH6`: -15.25 + 1042 = 1026.75

Implied offer size = min(1167, 8) = 8


The implied prices are the same in both cases, so the total implied size combines

Best implied bid price: 1026.25  size: 23 + 16 = 39

Best implied offer price: 1026.75  size: 46 + 8 = 54


5. Suppose a market order sells 5 `ZSF6`. At what price is the ordered filled? Is the fill potentially
an implied tade?

The best implied bid price matches the non-implied 1026.25 and will compete with the non-implied 72
bid size. The fill is potentially implied depending on the actual orders in the book.

6. Suppose a market order buys 5 `ZSF6`. At what price is the ordered filled? Is the fill potentially
an implied tade?

The best implied offer is 25 cents off the market, and there is sufficient size on the offer to
match with a 5 lot buy order, so there will be no implied trade.